### PR TITLE
add utility to back test Timer

### PR DESCRIPTION
**Pull request by robieta**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51123 Instruction count benchmark (7): Use  when collecting callgrind.
* #51122 Intermezzo: Add Timer support for callgrind repeats
* #51121 Intermezzo: quality of life tweaks to Timer. (Retain callgrind.out, change collect_baseline default, add __mul__ for FunctionCounts)
* #50610 Instruction count benchmark (6): Add frontend to drive microbenchmarks, and start README
* #50609 Instruction count benchmark (5): Add benchmark runner.
* #50608 Instruction count benchmark (4): Extend the benchmarks in both coverage and complexity.
* #50607 Instruction count benchmark (3): Enable TorchScript benchmarks
* #50606 Instruction count benchmark (2): Initial benchmark definitions.
* #50605 Instruction count benchmark (1): Types and interfaces
* #50604 add global_setup for C++ includes
* **#50603 add utility to back test Timer**

This PR is kind of dodgy, but given that I've tried to keep it as clean and localized as possible under the circumstances. In effect, it copies the local copy of `torch/utils` that `Timer` needs to a shadow folder in the target install. Where possible I've put the compat shims in `utils/benchmark` code, however some code mods are needed to support very old versions. I've tested `Timer` with Python snippets back to 0.4, and C++ snippets back to 1.0. This is code debt, it will break and have to be fixed as PyTorch gets updated; however it's only for a small set of devs and we can delete / prune as the historic analysis effort wraps up.

### CODE REVIEW

**Comment by ezyang** [comment]
This PR puts forth the hypothesis: "it's only for a small set of devs and we can delete / prune as the historic analysis effort wraps up."  I don't buy it. We're always going to be doing historical analysis--on a smaller time windows perhaps--but what is the WoS besides historical analysis. And so the more tightly integrated Timer is with PyTorch internals, the more likely it is to break when people make (seemingly unrelated) changes to functionality it relies on.

So, let me make the pitch from https://github.com/pytorch/pytorch/pull/44717#discussion_r492266813 again. It seems to me that there are two modes of use for timer:

1. To do benchmarking and analysis on master (or perhaps master plus a diff). You will always be running timer directly out of your PyTorch copy, it's OK to have as many features as you want and integrate as closely with PyTorch you want; all of this functionality can be tested the usual way with CI.
2. You want to do historical analysis, running a benchmark on old versions of PyTorch. You want **as little** dependence on PyTorch as possible; the more generic the library, the better (in fact, you probably should end up with something that is a generic benchmarking toolkit). When I did 0.4.1 comparison, it was literally one compiled C function I needed to patch in to do an analysis: 6c1c347fb16f792dbb2b9444c1a4bdd7fe98fe7#diff-889810736a005e292975703589788fd38d4b7778e152f9b28110f377d3de5a3f (`_valgrind_supported_platform` is in there but I didn't need it).

I understand what you are trying to do with this PR: you don't want to be in the business of maintaining two separate codebases for use cases (1) and (2). But it seems to me that you are getting the worst of both worlds: every time you add a new feature for (1) you have to carefully make sure it doesn't break backports, but the timer is still complicated enough that you need to go through big lengths to make it backport correctly (2).

**Comment by robieta** [comment]
> This PR puts forth the hypothesis: "it's only for a small set of devs and we can delete / prune as the historic analysis effort wraps up." I don't buy it. ...

Really the issue is `torch/utils/cpp_extension.py`; for everything else I can make `Timer` more or less version agnostic. (It's mostly just dealing with missing API symbols.) It seems to me that moving `Timer` out of tree won't solve the issue of "Timer expects modern PyTorch features even on older versions." I'm open to moving it to a separate repo/package, but it will add maintenance overhead so it will need to be carefully considered.

**Comment by ezyang** [comment]
For backtesting, how important do you think cpp_extension benchmarks are?

**Comment by robieta** [comment]
> For backtesting, how important do you think cpp_extension benchmarks are?

Pretty important, because that's what enables `Timer(..., language="c++")`, and deterministic runs for the historic sweep are pretty powerful. Of course, I could do a dumb Ninja/CMake compilation since I only use a small subset of `cpp_extension`, but then we're back to having two implementations of the same thing. (Maybe do that plus have a unit test that the Timer special case matches what one would get from `cpp_extension` in case it changes? It's suspect, but then so is this PR...)

**Comment by ezyang** [comment]
In 1:1 we were warming up to the idea of just having a really shitty ninja script for building C++ binaries for benchmarking.

**Comment by robieta** [comment]
@ezyang I hammered out a rough prototype for "just use a dumb specialized file" for C++ back testing. (And did some tidying since back-testing code was starting to seep into the rest of Timer.) It's not the most beautiful thing in the world, but it does seem reasonable and honestly seems less terrible and more transparent than a bunch of string rewrite rules on `cpp_extension`. (For normal use I think we still want to rely on `cpp_extension`, since it is more robust and thoroughly tested/developed.) WDYT?

File: torch/utils/benchmark/utils/historic/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+cmake_minimum_required(VERSION 3.18 FATAL_ERROR)
+
+project(timer_cpp_jit)
+set(CMAKE_CXX_STANDARD 14)
+
+message("name:            ${NAME}")
+message("torch_root:      ${TORCH_ROOT}")
+message("is_standalone:   ${IS_STANDALONE}")
+message("link_libtorch:   ${LINK_LIBTORCH}")
+message("valgrind_include ${VALGRIND_INCLUDE}")
```
Comment (ezyang) [inline torch/utils/benchmark/utils/historic/CMakeLists.txt:11]: You weren't kidding about "not most beautiful thing in the world" :P

Comment (ezyang) [reply torch/utils/benchmark/utils/historic/CMakeLists.txt:11]: You could make this marginally prettier by using `set()` to give these default values. Providing defaults can make it easier if people want to use this cmake to directly build code themselves or need to debug problems with the cmake.

Comment (robieta) [reply torch/utils/benchmark/utils/historic/CMakeLists.txt:11]: Would that help much? Bear in mind this is used for three different targets (depending on `${NAME}`), and most of the complexity is in the fiddly path magic that determines the values.

File: torch/utils/benchmark/utils/historic/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+else()
+    find_package(Python3 COMPONENTS Interpreter Development REQUIRED)
+    find_package(PythonInterp REQUIRED)  # Without this PyBind11 will find the wrong version
+    find_package(pybind11 CONFIG REQUIRED)
+    add_compile_definitions(TORCH_EXTENSION_NAME=${NAME})
+    pybind11_add_module(${NAME} "${PROJECT_SOURCE_DIR}/${NAME}.cpp")
+endif()
+
+# Wall time doesn't need this, but it doesn't hurt.
+target_include_directories(${NAME} PRIVATE ${VALGRIND_INCLUDE})
```
Comment (ezyang) [inline torch/utils/benchmark/utils/historic/CMakeLists.txt:24]: This line is a bit surprising for two reasons: (1) at least whenever I've used it, Valgrind was dunked in the standard /usr/include so I didn't actually need to do anything to include the header, and (2) even if it wasn't, typical operating procedure in cmake files is to use `find_package` to find the appropriate includes. Curious why you needed it like this here.

File: torch/utils/benchmark/utils/historic/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+message("is_standalone:   ${IS_STANDALONE}")
+message("link_libtorch:   ${LINK_LIBTORCH}")
+message("valgrind_include ${VALGRIND_INCLUDE}")
+
+if(${IS_STANDALONE})
+    add_executable(${NAME} "${PROJECT_SOURCE_DIR}/${NAME}.cpp")
+else()
+    find_package(Python3 COMPONENTS Interpreter Development REQUIRED)
+    find_package(PythonInterp REQUIRED)  # Without this PyBind11 will find the wrong version
+    find_package(pybind11 CONFIG REQUIRED)
```
Comment (ezyang) [inline torch/utils/benchmark/utils/historic/CMakeLists.txt:18]: Is the intention for this to find the pybind11 we have vendored in third party?

File: torch/utils/benchmark/utils/historic/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    add_compile_definitions(TORCH_EXTENSION_NAME=${NAME})
+    pybind11_add_module(${NAME} "${PROJECT_SOURCE_DIR}/${NAME}.cpp")
+endif()
+
+# Wall time doesn't need this, but it doesn't hurt.
+target_include_directories(${NAME} PRIVATE ${VALGRIND_INCLUDE})
+
+if(${LINK_LIBTORCH})
```
Comment (ezyang) [inline torch/utils/benchmark/utils/historic/CMakeLists.txt:26]: When would you *not* want to link libtorch?

File: torch/utils/benchmark/utils/historic/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    list(APPEND CMAKE_PREFIX_PATH ${TORCH_ROOT})
+    find_package(Torch REQUIRED)
+    find_package(Threads REQUIRED)
+
+    target_include_directories(${NAME} PRIVATE ${TORCH_INCLUDE_DIRS})
+
+    # Older versions of PyTorch struggle to link MKL.
```
Comment (ezyang) [inline torch/utils/benchmark/utils/historic/CMakeLists.txt:33]: I'm not sure how this comment relates to the code below

File: torch/utils/benchmark/utils/cpp_jit.py [hunk 6]
```diff
@@ -11,10 +11,14 @@
 
 import torch
```
Comment (ezyang) [inline torch/utils/benchmark/utils/cpp_jit.py:12]: I'm not sure what's going on with the fuzz in this file

File: torch/utils/benchmark/utils/historic/patch.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
+"""Utility to patch benchmark utils into earlier versions of PyTorch.
```
Comment (ezyang) [inline torch/utils/benchmark/utils/historic/patch.py:1]: This file is dead now, with the new approach?

File: torch/utils/benchmark/utils/historic/CMakeLists.txt (hunk 2 above)
Comment (ezyang) [reply torch/utils/benchmark/utils/historic/CMakeLists.txt:24]: Answer: this is pointing to the valgrind wrapper directory, not the system valgrind installation.

**Comment by ezyang** [comment]
The new `compat_jit` script looks pretty short and pretty maintainable. Are you going to publish a smoketest script for testing that it works with N recent release of PyTorch? (I'm guessing that's the manual testing you did). There are still a bunch of other changes in this PR that I am not sure what is going on with them.

**Pull request closed without merging**